protocols/dcutr/src/protocol/outbound.rs

```diff
@@ -0,0 +1,164 @@
+use thiserror::Error;
+use unsigned_varint::codec::UviBytes;
+
+pub struct Upgrade {
+    obs_addrs: Vec<Multiaddr>,
+}
+
+impl upgrade::UpgradeInfo for Upgrade {
+    type Info = &'static [u8];
+    type InfoIter = iter::Once<Self::Info>;
+
+    fn protocol_info(&self) -> Self::InfoIter {
+        iter::once(super::PROTOCOL_NAME)
+    }
+}
+
+impl Upgrade {
+    pub fn new(obs_addrs: Vec<Multiaddr>) -> Self {
+        Self { obs_addrs }
+    }
+}
+
+impl upgrade::OutboundUpgrade<NegotiatedSubstream> for Upgrade {
+    type Output = Connect;
+    type Error = UpgradeError;
+    type Future = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+
+    fn upgrade_outbound(self, substream: NegotiatedSubstream, _: Self::Info) -> Self::Future {
+        let msg = HolePunch {
+            r#type: hole_punch::Type::Connect.into(),
+            obs_addrs: self.obs_addrs.into_iter().map(|a| a.to_vec()).collect(),
+        };
+
+        let mut encoded_msg = BytesMut::new();
+        msg.encode(&mut encoded_msg)
+            .expect("BytesMut to have sufficient capacity.");
+
+        let mut codec = UviBytes::default();
+        codec.set_max_len(super::MAX_MESSAGE_SIZE_BYTES);
+        let mut substream = Framed::new(substream, codec);
+
+        async move {
+            substream.send(encoded_msg.freeze()).await?;
+
+            let sent_time = Instant::now();
+
+            let msg: bytes::BytesMut = substream
+                .next()
+                .await
+                .ok_or(std::io::Error::new(std::io::ErrorKind::UnexpectedEof, ""))??;
+
+            let rtt = sent_time.elapsed();
+
+            let HolePunch { r#type, obs_addrs } = HolePunch::decode(Cursor::new(msg))?;
+
+            let r#type = hole_punch::Type::from_i32(r#type).ok_or(UpgradeError::ParseTypeField)?;
+            match r#type {
+                hole_punch::Type::Connect => {}
+                hole_punch::Type::Sync => return Err(UpgradeError::UnexpectedTypeSync),
+            }
+
+            let obs_addrs = if obs_addrs.is_empty() {
+                return Err(UpgradeError::NoAddresses);
+            } else {
+                obs_addrs
+                    .into_iter()
+                    .map(Multiaddr::try_from)
+                    // Filter out relayed addresses.
+                    .filter(|a| match a {
+                        Ok(a) => !a.iter().any(|p| p == Protocol::P2pCircuit),
+                        Err(_) => true,
+                    })
+                    .collect::<Result<Vec<Multiaddr>, _>>()
+                    .map_err(|_| UpgradeError::InvalidAddrs)?
+            };
+
+            let msg = HolePunch {
+                r#type: hole_punch::Type::Sync.into(),
+                obs_addrs: vec![],
+            };
+
+            let mut encoded_msg = BytesMut::new();
+            msg.encode(&mut encoded_msg)
+                .expect("BytesMut to have sufficient capacity.");
+
+            substream.send(encoded_msg.freeze()).await?;
```

Comment:
Might be worthwhile to extract some helpers to hide the details of constructing one of these messages from the overall flow.

Reply:
Good idea. We can even go one step further and implement `asynchronous_codec::Encoder` and `asynchronous_codec::Decoder` for a custom `Codec` that takes care of all the `prost::Message` encoding, wrapping a `UviBytes`.

What do you think of 354900fd1eafa108d9f70f54ec5c15ae3527656b @thomaseizinger?

We could take this one step further, making `Codec` abstract over `prost::Message` and thus use the same `Codec` in all `protocols/XXX`. What do you think?